### PR TITLE
Support MQTT on LWIP Ethernet boards

### PR DESCRIPTION
**Pull request by rechrtb**
- Support outgoing connections and MQTT on Ethernet boards using Lwip
- Add handling of enabling/disabling MQTT on boards with multiple network interfaces; only one active MQTT instance is supported, has to be disabled on the previously disabled interface first (if any) using `M586 P4 S0 I<num>`. See notes about in https://github.com/Duet3D/MQTT-WPA2-Enterprise-Demo/tree/ethernet_mqtt/mqtt#on-boards-with-multiple-network-interfaces
- Add support for specifying clean (or not) session; uses boolean parameter 'N' for `M586.4`.

Related PR: https://github.com/Duet3D/MQTT-WPA2-Enterprise-Demo/pull/2

### CODE REVIEW

File: src/GCodes/GCodes2.cpp [hunk 1]
```diff
@@ -3921,20 +3923,52 @@ bool GCodes::HandleMcode(GCodeBuffer& gb, const StringRef& reply) THROWS(GCodeEx
 
 											IPAddress ip;
 
-											if (protocol == MqttProtocol)
+#if SUPPORT_MQTT
+											if (interface < reprap.GetNetwork().GetNumNetworkInterfaces() && protocol == MqttProtocol)
```
Comment (dc42) [inline src/GCodes/GCodes2.cpp]: This code is getting too long to be included in GCodes2.cpp. I suggest you move all of the code after "if (CheckNetworkCommandAllowed(gb, reply, result)" into a new function in Network.cpp, just as we do for M587.

File: src/Networking/NetworkClient.cpp [hunk 2]
```diff
@@ -12,49 +12,27 @@
...
 
-bool NetworkClient::Start(NetworkProtocol protocol, NetworkInterface *iface) noexcept
+bool NetworkClient::Start(NetworkProtocol protocol, NetworkInterface *interface) noexcept
 {
 	if (HandlesProtocol(protocol))
 	{
 		if (skt == nullptr)
 		{
-			if (interface != nullptr)
+			if (millis() - whenRequest < connectTimeout)
```
Comment (dc42) [inline src/Networking/NetworkClient.cpp:26]: You have removed the check for interface != nullptr here. Are you certain that this function can't be called with interface == nullptr and skt != nullptr? For example, what will happen on a 6HC without a wifi module attached if you try to start MQTT on interface #1 ? Or is it the case that interface == nullptr is actually OK here?

Comment (rechrtb) [reply src/Networking/NetworkClient.cpp:26]: The `interface` here was a class member that has been supplanted by keeping track of the interface number instead. 

Furthermore, this check is supplanted by checking in `MqttClient.cpp`. 

Lastly, this code is called only when the specific `NetworkInterface` is in the `NetworkState::connected`. I double checked this, and if the wifi module is not attached, the network state remains at `NetworkState::disabled` and so this code never gets called, even when MQTT is enabled on the wifi interface.

Comment (dc42) [reply src/Networking/NetworkClient.cpp:26]: Thanks for the clarification..

File: src/Networking/Network.cpp [hunk 3]
```diff
@@ -565,6 +544,121 @@ void Network::Exit() noexcept
 #endif // HAS_NETWORKING
 }
 
+#if HAS_NETWORKING
+GCodeResult Network::ConfigureNetworkProtocol(GCodeBuffer& gb, const StringRef& reply)
+{
```
Comment (dc42) [inline src/Networking/Network.cpp:549]: There are several redundant calls to reprap.GetNetwork() in this function. Please remove them.

**Pull request merged**